### PR TITLE
swarm/pss: pssclient, pss tests

### PR DESCRIPTION
**Pull request by nolash**
reinstated full sim pss core tests
pssclient; p2p/protocols mountable rpc proxy for pss ws)

### CODE REVIEW

File: swarm/pss/client/client.go [hunk 1]
```diff
@@ -9,58 +9,77 @@ import (
...
+
 type PssClient struct {
 	localuri     string
 	remoteuri    string
 	ctx          context.Context
 	cancel       func()
 	subscription *rpc.ClientSubscription
 	topicsC      chan []byte
-	msgC         chan PssAPIMsg
+	msgC         chan pss.PssAPIMsg
```
Comment (zelig) [inline swarm/pss/client/client.go]: this should be pointer `*pss.PssAPIMsg` no?

File: swarm/pss/client/client.go [hunk 2]
```diff
@@ -9,58 +9,77 @@ import (
...
 	quitC        chan struct{}
 	quitting     uint32
 	ws           *rpc.Client
 	lock         sync.Mutex
-	peerPool     map[PssTopic]map[pot.Address]*pssRPCRW
-	protos       []*p2p.Protocol
+	peerPool     map[pss.PssTopic]map[pot.Address]*pssRPCRW
+	protos       map[pss.PssTopic]*p2p.Protocol
```
Comment (zelig) [inline swarm/pss/client/client.go]: as suggested on #100 let;s get rid of the Pss prefixes, will make the code more readable
eg. `pss.Topic, pss.API`

File: swarm/pss/client/client.go [hunk 3]
```diff
@@ -69,72 +88,84 @@ func (rw *pssRPCRW) ReadMsg() (p2p.Msg, error) {
...
-		peerPool: make(map[PssTopic]map[pot.Address]*pssRPCRW),
+		peerPool: make(map[pss.PssTopic]map[pot.Address]*pssRPCRW),
+		protos:   make(map[pss.PssTopic]*p2p.Protocol),
 		ctx:      ctx,
 		cancel:   cancel,
 	}
 
-	if remotehost == "" {
-		remotehost = "localhost"
+	if config.RemoteHost == "" {
```
Comment (zelig) [inline swarm/pss/client/client.go]: use defaultconfig and overwrite

File: swarm/pss/client/client.go [hunk 4]
```diff
@@ -69,72 +88,84 @@ func (rw *pssRPCRW) ReadMsg() (p2p.Msg, error) {
...
 
-	pssc.remoteuri = fmt.Sprintf("%s://%s:%d", prefix, remotehost, remoteport)
-	pssc.localuri = fmt.Sprintf("%s://%s", prefix, originhost)
+	pssc.remoteuri = fmt.Sprintf("%s://%s:%d", prefix, config.RemoteHost, config.RemotePort)
+	pssc.localuri = fmt.Sprintf("%s://%s", prefix, config.SelfHost)
 
 	return pssc
 }
 
+func NewPssClientWithRPC(ctx context.Context, client *rpc.Client) *PssClient {
```
Comment (zelig) [inline swarm/pss/client/client.go]: drop Pss in the names everywhere

File: swarm/pss/client/client.go [hunk 5]
```diff
@@ -69,72 +88,84 @@ func (rw *pssRPCRW) ReadMsg() (p2p.Msg, error) {
...
+		quitC:    make(chan struct{}),
+		peerPool: make(map[pss.PssTopic]map[pot.Address]*pssRPCRW),
+		protos:   make(map[pss.PssTopic]*p2p.Protocol),
+		ws:       client,
+		ctx:      ctx,
+	}
+}
+
 func (self *PssClient) shutdown() {
 	atomic.StoreUint32(&self.quitting, 1)
```
Comment (zelig) [inline swarm/pss/client/client.go]: this line is inert AFAIC who checks the quitting bool var?

File: swarm/pss/client/client_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+	go func() {
+		<-ping.QuitC
+		client.cancel()
+	}()
+
+	select {
+	case <-client.ctx.Done():
+		t.Fatalf("outgoing timed out or canceled")
+	default:
```
Comment (zelig) [inline swarm/pss/client/client_test.go]: default casefalls through, the test can never fail, same for outgoing.
you need select here between ping.QuitC and ctx no?

File: swarm/pss/client/client_test.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	ps := pss.NewTestPss(nil)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	var addr []byte
+	ping := &pss.PssPing{
+		QuitC: make(chan struct{}),
+	}
+	proto := newProtocol(ping)
+	client, err := baseTester(t, proto, ps, ctx, cancel, quitC)
+	if err != nil {
+		t.Fatalf(err.Error())
```
Comment (zelig) [inline swarm/pss/client/client_test.go:51]: `t.Fatalf(err)`

File: swarm/pss/client/client_test.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	client.ws.Call(&addr, "psstest_baseAddr")
+
+	code, _ := pss.PssPingProtocol.GetCode(&pss.PssPingMsg{})
+	rlpbundle, err := pss.NewProtocolMsg(code, &pss.PssPingMsg{
+		Created: time.Now(),
+	})
+	if err != nil {
+		t.Fatalf("couldn't make pssmsg")
```
Comment (zelig) [inline swarm/pss/client/client_test.go]: add the actual err to fatalf

File: swarm/pss/common.go [hunk 9]
```diff
@@ -11,115 +11,95 @@ import (
...
-	quitC chan struct{}
+type PssPing struct {
+	QuitC chan struct{}
 }
 
-func (self *pssPing) pssPingHandler(msg interface{}) error {
+func (self *PssPing) PssPingHandler(msg interface{}) error {
 	log.Warn("got ping", "msg", msg)
-	self.quitC <- struct{}{}
+	self.QuitC <- struct{}{}
```
Comment (zelig) [inline swarm/pss/common.go]: really confusing to call this `QuitC`, nothing quits

File: swarm/pss/pss.go [hunk 10]
```diff
@@ -2,19 +2,18 @@ package pss
 
 import (
 	"bytes"
-	"encoding/binary"
+	//"encoding/binary"
```
Comment (zelig) [inline swarm/pss/pss.go]: remove commented lines

File: swarm/pss/pss.go [hunk 11]
```diff
@@ -341,7 +280,7 @@ func (self *Pss) Process(pssmsg *PssMsg) error {
 	return nil
 }
 
-// Sends a message using pss. The message could be anything at all, and will be handled by whichever handler function is mapped to PssTopic using *Pss.Register()
+// Sends a message using  The message could be anything at all, and will be handled by whichever handler function is mapped to PssTopic using *Pss.Register()
 //
 // The to address is a swarm overlay address
 func (self *Pss) Send(to []byte, topic PssTopic, msg []byte) error {
```
Comment (zelig) [inline swarm/pss/pss.go]: either calll this SendRaw, or do not embed pss in api

File: swarm/pss/pss.go [hunk 12]
```diff
@@ -446,6 +385,10 @@ func (self *Pss) removePeerTopic(rw p2p.MsgReadWriter, topic PssTopic) {
 	}
 }
 
+func (self *Pss) isSelfRecipient(msg *PssMsg) bool {
+	return bytes.Equal(msg.To, self.Overlay.BaseAddr())
+}
+
 func (self *Pss) isActive(id pot.Address, topic PssTopic) bool {
 	return self.peerPool[id][topic] != nil
```
Comment (zelig) [inline swarm/pss/pss.go]: what if id not a key.? this will panic

File: swarm/pss/pss.go [hunk 13]
```diff
@@ -505,20 +448,20 @@ type PssProtocol struct {
...
 	}
-	pss.Register(topic, pp.handle)
+	ps.Register(topic, pp.handle)
 	return nil
 }
 
 func (self *PssProtocol) handle(msg []byte, p *p2p.Peer, senderAddr []byte) error {
 	hashoaddr := pot.NewHashAddressFromBytes(senderAddr).Address
-	if !self.isActive(hashoaddr, *self.topic) {
+	if !self.Pss.isActive(hashoaddr, *self.topic) {
```
Comment (zelig) [inline swarm/pss/pss.go]: do not need Pss, just call isActive directly

File: swarm/pss/pssapi.go [hunk 14]
```diff
@@ -1,12 +1,14 @@
...
 
 import (
+	"bytes"
 	"context"
 	"fmt"
 
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/p2p"
 	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/network"
```
Comment (zelig) [inline swarm/pss/pssapi.go:11]: do not emped Pss or call Pss.Send something else

File: swarm/pss/pssapi.go [hunk 15]
```diff
@@ -19,15 +21,8 @@ func NewPssAPI(ps *Pss) *PssAPI {
...
-// PssAPIMsg is the type for messages, it extends the rlp encoded protocol Msg
-// with the Sender's overlay address
-type PssAPIMsg struct {
-	Msg  []byte
-	Addr []byte
-}
-
 // NewMsg API endpoint creates an RPC subscription
-func (pssapi *PssAPI) NewMsg(ctx context.Context, topic PssTopic) (*rpc.Subscription, error) {
+func (pssapi *PssAPI) ReceivePss(ctx context.Context, topic PssTopic) (*rpc.Subscription, error) {
```
Comment (zelig) [inline swarm/pss/pssapi.go]: lets stick with the most intuitive simple names for the API: Send and Receive

File: swarm/pss/pssapi.go [hunk 16]
```diff
@@ -61,16 +55,49 @@ func (pssapi *PssAPI) NewMsg(ctx context.Context, topic PssTopic) (*rpc.Subscrip
...
+	if pssapi.Pss.debug && bytes.Equal(msg.Addr, pssapi.Pss.BaseAddr()) {
+		log.Warn("Pss debug enabled; send to self shortcircuit", "apimsg", msg, "topic", topic)
+		env := NewPssEnvelope(msg.Addr, topic, msg.Msg)
+		return pssapi.Pss.Process(&PssMsg{
+			To: pssapi.Pss.BaseAddr(),
+			Payload: env,
+		})
+	}
+	return pssapi.Pss.Send(msg.Addr, topic, msg.Msg)
+	/*if err != nil {
```
Comment (zelig) [inline swarm/pss/pssapi.go]: remove comments

File: swarm/pss/simulations/service_test.go [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+	return []rpc.API{{
+		Namespace: "test",
+		Version:   "1.0",
+		Service:   &TestAPI{t.pss, t.handshakes},
+	}}
+}
+
+func (t *testService) Start(*p2p.Server) error {
+	return t.pss.RunProtocol(&p2p.Protocol{
+		Name:    "test",
```
Comment (zelig) [inline swarm/pss/simulations/service_test.go:175]: this should be in Protocols no?

File: swarm/pss/client/client_test.go [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	pssenv := pss.PssEnvelope{
```
Comment (zelig) [inline swarm/pss/client/client_test.go]: use NewPssEnvelope no?

**Review by zelig — CHANGES_REQUESTED** [review]
i have not yet checked pss_test.go

File: swarm/pss/common.go (hunk 9 above)
Comment (nolash) [reply swarm/pss/common.go]: I'll change it to ping.C

File: swarm/pss/pss.go (hunk 11 above)
Comment (nolash) [reply swarm/pss/pss.go]: You mean in order to be able to use Send method name in PssAPI? Is it a bug that it doesn't override it, by the way?

File: swarm/pss/simulations/service_test.go (hunk 17 above)
Comment (nolash) [reply swarm/pss/simulations/service_test.go:175]: This test is half-baked still, can we just leave it as is please?

File: swarm/pss/client/client.go (hunk 1 above)
Comment (nolash) [reply swarm/pss/client/client.go]: I don't think so, it's the channel listening to subscribes from RPC. If it was wrong it wouldn't work.?

File: swarm/pss/client/client_test.go (hunk 7 above)
Comment (nolash) [reply swarm/pss/client/client_test.go:51]: gives error...

**Comment by nolash** [comment]
Maybe I suggest that since the pss_test.go tests run but don't work 100% anyway, let's just merge this without too much nitpicking, then do another pass at getting them up and running with the revised kademlia?

**Pull request merged**